File: windows/Pcci/Services/Preferences.cs
using System;
using System.IO;
using System.Text.Json;
using System.Text.Json.Serialization;

namespace Pcci.Services;

/// <summary>
/// What the app remembers between launches.
///
/// A small JSON file beside the engine's own cache and logs. Nothing here is worth an
/// error message: a missing or unreadable file means "use the defaults", and a failed
/// write means the next launch starts from the defaults again. Neither is worth
/// interrupting somebody setting up for a Sunday.
/// </summary>
public sealed class Preferences
{
    [JsonPropertyName("export_target")] public string ExportTarget { get; set; } = "propresenter";
    [JsonPropertyName("chord_delivery")] public string ChordDelivery { get; set; } = "inline";
    [JsonPropertyName("chord_placement")] public string ChordPlacement { get; set; } = "chords_only";
    [JsonPropertyName("chords_on_slide")] public bool ChordsOnSlide { get; set; }
    [JsonPropertyName("write_chordpro")] public bool WriteChordPro { get; set; }
    [JsonPropertyName("lines_per_slide")] public int LinesPerSlide { get; set; } = 4;

    /// <summary>
    /// Whether an export stops to show these first. On until somebody says otherwise,
    /// because the settings decide what reaches a stage screen and the first export is
    /// the moment to find that out - not the Sunday after.
    /// </summary>
    [JsonPropertyName("ask_before_export")] public bool AskBeforeExport { get; set; } = true;

    public static string FilePath
    {
        get
        {
            var root = Environment.GetEnvironmentVariable("LOCALAPPDATA");
            if (string.IsNullOrEmpty(root))
            {
                root = Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData);
            }
            return Path.Combine(root, "PCCI", "settings.json");
        }
    }

    public static Preferences Load()
    {
        try
        {
            if (!File.Exists(FilePath)) return new Preferences();
            var text = File.ReadAllText(FilePath);
            return JsonSerializer.Deserialize<Preferences>(text) ?? new Preferences();
        }
        catch (Exception)
        {
            return new Preferences();
        }
    }

    public void Save()
    {
        try
        {
            var directory = Path.GetDirectoryName(FilePath);
            if (!string.IsNullOrEmpty(directory)) Directory.CreateDirectory(directory);
            var options = new JsonSerializerOptions { WriteIndented = true };
            // Through a temporary file: an interrupted write must not leave a settings
            // file that cannot be read, which would silently reset everything.
            var temporary = FilePath + ".tmp";
            File.WriteAllText(temporary, JsonSerializer.Serialize(this, options));
            File.Move(temporary, FilePath, overwrite: true);
        }
        catch (Exception)
        {
            // Settings that do not survive a restart are a nuisance, not a failure.
        }
    }
}

File: macos/Sources/PCCI/Preferences.swift
import Foundation

/// What the app remembers between launches.
///
/// `UserDefaults`, because that is where a Mac app's settings belong and the whole of
/// it is seven small values. Nothing here is worth an error message: a key that is
/// missing or holds something unexpected means "use the default", which is also what a
/// first launch sees.
struct Preferences {
    var exportTarget: ExportTarget = .propresenter
    var chordDelivery: ChordDelivery = .inline
    var chordPlacement: ChordPlacementStyle = .chordsOnly
    var chordsOnSlide = false
    var writeChordPro = false
    var linesPerSlide = 4

    /// Whether an export stops to show the settings first. On until somebody says
    /// otherwise: these settings decide what reaches a stage screen, and an export is
    /// the moment to find that out rather than the Sunday after.
    var askBeforeExport = true

    private enum Key {
        static let exportTarget = "exportTarget"
        static let chordDelivery = "chordDelivery"
        static let chordPlacement = "chordPlacement"
        static let chordsOnSlide = "chordsOnSlide"
        static let writeChordPro = "writeChordPro"
        static let linesPerSlide = "linesPerSlide"
        static let askBeforeExport = "askBeforeExport"
    }

    static func load(from defaults: UserDefaults = .standard) -> Preferences {
        var preferences = Preferences()
        if let raw = defaults.string(forKey: Key.exportTarget),
            let value = ExportTarget(rawValue: raw)
        {
            preferences.exportTarget = value
        }
        if let raw = defaults.string(forKey: Key.chordDelivery),
            let value = ChordDelivery(rawValue: raw)
        {
            preferences.chordDelivery = value
        }
        if let raw = defaults.string(forKey: Key.chordPlacement),
            let value = ChordPlacementStyle(rawValue: raw)
        {
            preferences.chordPlacement = value
        }
        preferences.chordsOnSlide = defaults.bool(forKey: Key.chordsOnSlide)
        preferences.writeChordPro = defaults.bool(forKey: Key.writeChordPro)
        // A missing integer reads as zero, which is not a legal number of lines.
        let lines = defaults.integer(forKey: Key.linesPerSlide)
        preferences.linesPerSlide = (1...10).contains(lines) ? lines : 4
        if defaults.object(forKey: Key.askBeforeExport) != nil {
            preferences.askBeforeExport = defaults.bool(forKey: Key.askBeforeExport)
        }
        return preferences
    }

    func save(to defaults: UserDefaults = .standard) {
        defaults.set(exportTarget.rawValue, forKey: Key.exportTarget)
        defaults.set(chordDelivery.rawValue, forKey: Key.chordDelivery)
        defaults.set(chordPlacement.rawValue, forKey: Key.chordPlacement)
        defaults.set(chordsOnSlide, forKey: Key.chordsOnSlide)
        defaults.set(writeChordPro, forKey: Key.writeChordPro)
        defaults.set(linesPerSlide, forKey: Key.linesPerSlide)
        defaults.set(askBeforeExport, forKey: Key.askBeforeExport)
    }
}
